go-controller/pkg/ovn/controller/apbroute/external_controller_policy.go

```diff
@@ -59,9 +62,21 @@ func (m *externalPolicyManager) applyProcessedPolicy(policyName string, routePol
 	return nil
 }
 
-// processDeletePolicy takes in a policy and proceeds to delete the gateway IPs derived from the static and dynamic hops from the namespaces impacted by the policy, as defined by the namespace
-// selector in the from field.
+// processDeletePolicy takes in a policy, marks it for deletion and proceeds to delete the gateway IPs derived from the static and dynamic hops from the namespaces impacted by the policy, as defined by the namespace
+// selector in the from field. The last step is to delete it from the cache.
 func (m *externalPolicyManager) processDeletePolicy(routePolicy *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute) error {
+	_, found, markedForDeletion := m.getRoutePolicyFromCache(routePolicy.Name)
```

Comment:
you could do lines 68-78 all in one DoWithLock. GetAndMarkForDeletion(routePolicy.Name). If you find it, mark for deletion and return it. Otherwise return nil.